homework-submission/homework-week1/app.js

```diff
@@ -0,0 +1,160 @@
+const Item = ({ id, description, deadline, done, removeItem }) => {
+  return (
+    <li key={id} className={done === true ? 'true' : ''}>
+      {description} {deadline} {done}
+      <span className="delete" onClick={() => removeItem(id)}>
+        DELETE
+      </span>
+    </li>
+  );
+};
+
+const TodoItems = ({ toDoList, removeItem }) => {
+  const length = toDoList.length;
+  const ListItems = length ? (
+    toDoList.map(({ id, done, description, deadline }) => {
+      return (
+        <Item
+          key={id}
+          id={id}
+          description={description}
+          deadline={deadline}
+          done={done}
+          removeItem={removeItem}
+        />
+      );
+    })
+  ) : (
+    <p>There is no items</p>
+  );
+
+  return <ul className="todo-items">{ListItems}</ul>;
+};
+
+class AddItems extends React.Component {
+  state = {
+    description: '',
+    deadline: '',
+  };
+
+  handelChange = e => {
+    this.setState({
+      [e.target.id]: e.target.value,
+    });
+  };
+
+  error() {
+    let element = document.getElementById('error');
+    element.classList.remove('hidde');
+  }
+
+  handelSubmit = e => {
+    e.preventDefault();
+    if (e.target['description'].value !== '' && e.target['deadline'].value !== '') {
+      this.props.addItem(this.state);
+      this.setState({
+        description: '',
+        deadline: '',
+      });
+    } else {
+      this.error();
+    }
+  };
+
+  render() {
+    return (
+      <div>
+        <form onSubmit={this.handelSubmit}>
+          <label>Description: </label>
+          <input
+            type="text"
+            id="description"
+            value={this.state.description}
+            placeholder="Enter a short description here..."
+            onChange={this.handelChange}
+          />
+
+          <label>Deadline: </label>
+          <input
+            type="date"
+            id="deadline"
+            placeholder="Enter deadline..."
+            value={this.state.deadline}
+            onChange={this.handelChange}
+          />
+          <input type="submit" value="Add" className="add" />
+        </form>
+      </div>
+    );
+  }
+}
+
+class App extends React.Component {
+  state = {
+    toDoList: [
+      {
+        id: 1,
+        description: 'Get out of bed',
+        deadline: '2017-09-11',
+        done: true,
+      },
+      {
+        id: 2,
+
+        description: 'Brush teeth',
+        deadline: '2017-09-10',
+        done: false,
+      },
+      {
+        id: 3,
+        description: 'Eat breakfast',
+        deadline: '2017-09-09',
+        done: false,
+      },
+    ],
+  };
+
+  removeItem = id => {
+    const toDoList = this.state.toDoList.filter(item => {
+      return item.id !== id;
+    });
+    this.setState({ toDoList });
+  };
+
+  addItem = item => {
+    const random = Math.random(Math.random() * 9999);
+    item.id = random;
+    let items = this.state.toDoList;
+    items.push(item);
+    this.setState({ items });
```

Comment:
if you tried the following you will notice a direct change of the state
```javascript
    console.log(this.state.toDoList.length) // if this 3
    items.push(item); // this line is mutating the original state here!
    console.log(this.state.toDoList.length) // this is 4 and it should stay 3
```

This direct change shouldn't be happening, the main reason of `.setState` is making things asynchronous.

the `.setState` method isn't only for rendering the content, it's made on top of a complete life cycle happening on behalf of your program, and we want to make sure we are not changing the state by hand as here.

the best solution here would be as the following:
```javascript
    this.setState({ items: [ ...items, item ] }); // and that how easy it's 😃
```